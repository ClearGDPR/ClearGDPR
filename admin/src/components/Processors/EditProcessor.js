import React from 'react';
import PropTypes from 'prop-types';
import { Form, TextArea, Checkbox } from 'react-form';

import TextInput from '../core/cards/dashboard/TextInput';
import Loader from '../core/cards/dashboard/Loader';

// TODO: this should be fetch from configuration service
const DEMO_SCOPES = {
  'Full Name': 'user.fullName',
  Email: 'user.email',
  'Phone Number': 'user.phoneNumber'
};

export class EditProcessor extends React.Component {
  static propTypes = {
    isLoading: PropTypes.bool,
    onSubmit: PropTypes.func,
    touched: PropTypes.object,
    errors: PropTypes.object
  };

  renderForm() {
    return (
      <React.Fragment>
        <TextInput
          label="Name"
          placeholder="*********"
          type={'text'}
          error={
            this.props.touched &&
            this.props.touched['name'] &&
            this.props.errors &&
            this.props.errors['name']
          }
          field="name"
          required
        />

        <label htmlFor="description">Description:</label>
        <TextArea field="description" id="description" />

        <TextInput
          label="Logo URL"
          placeholder="*********"
          type={'text'}
          error={
            this.props.touched &&
            this.props.touched['logoUrl'] &&
            this.props.errors &&
            this.props.errors['logoUrl']
          }
          field="logoUrl"
          required
        />

        <label>Scopes:</label>
        {Object.keys(DEMO_SCOPES).map((scope, i) => <Checkbox key={i} field={`scope[${scope}]`} />)}

        <button type="submit" className="btn">
          Submit
        </button>
      </React.Fragment>
    );
  }

  render() {
    return (
      <form onSubmit={this.props.onSubmit}>
        {this.props.isLoading ? <Loader /> : this.renderForm()}
      </form>
    );
  }
}

const EditProcessorForm = props => (
  <Form onSubmit={submittedValues => props.onSubmit && props.onSubmit(submittedValues)}>
    {formApi => (
      <EditProcessor
        {...props}
        onSubmit={formApi.submitForm}
        errors={{ ...formApi.errors, ...props.errors }}
        touched={formApi.touched}
      />
    )}
  </Form>
);

EditProcessorForm.propTypes = {
  onSubmit: PropTypes.func,
  errors: PropTypes.object
};

export default EditProcessorForm;
