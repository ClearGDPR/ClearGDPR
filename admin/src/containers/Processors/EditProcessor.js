import React from 'react';
import PropTypes from 'prop-types';

import { ProcessorsConsumer } from './ProcessorsContext';
import { PanelConsumer } from 'containers/MainLayout/PanelContext';
import EditProcessorForm from 'components/Processors/EditProcessor';

export class EditProcessorContainer extends React.Component {
  static propTypes = {
    updateProcessor: PropTypes.func,
    processor: PropTypes.object,
    closePanel: PropTypes.func
  };

  state = {
    errors: {}
  };

  onSubmit(processor) {
    // Clone data to avoid updating form until save is done
    const processorData = Object.assign({}, processor);
    delete processorData.address;
    processorData.scopes = Object.keys(processor.scopes).reduce((scopes, s) => {
      return processor.scopes[s] ? scopes.concat(s) : scopes;
    }, []);

    return this.props
      .updateProcessor(processorData)
      .then(() => this.props.closePanel())
      .catch(e =>
        this.setState({
          errors: {
            processor: e.toString()
          }
        })
      );
  }

  render() {
    return (
      <EditProcessorForm
        values={this.props.processor}
        errors={this.state.errors}
        onSubmit={this.onSubmit.bind(this)}
        isLoading={this.state.isLoading}
      />
    );
  }
}

export default props => (
  <PanelConsumer>
    {({ closePanel }) => (
      <ProcessorsConsumer>
        {({ updateProcessor, isLoading }) => (
          <EditProcessorContainer
            {...props}
            isLoading={isLoading}
            updateProcessor={updateProcessor}
            closePanel={closePanel}
          />
        )}
      </ProcessorsConsumer>
    )}
  </PanelConsumer>
);
