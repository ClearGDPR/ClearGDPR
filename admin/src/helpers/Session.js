import cookie from 'cookie';

export class Session {
  set(authResult) {
    const maxAge = 1 * 24 * 60 * 60; // 1 day long;
    document.cookie = cookie.serialize('auth', authResult, { maxAge });
    document.cookie = cookie.serialize('expires_at', authResult, { maxAge });
  }

  destroy() {
    document.cookie = cookie.serialize('auth', '', { maxAge: -1 });
    document.cookie = cookie.serialize('expires_at', '', { maxAge: -1 });
  }

  isLoggedIn() {
    let expiresAt = cookie.parse(document.cookie).expires_at || '';
    return new Date().getTime() < expiresAt;
  }
}

export default new Session();
