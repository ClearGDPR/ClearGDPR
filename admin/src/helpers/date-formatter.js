import { format } from 'date-fns';

export default date => {
  return format(new Date(date), 'DD/MM/YYYY h:mma');
};
