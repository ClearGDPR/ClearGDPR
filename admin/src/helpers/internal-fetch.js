import merge from 'lodash/object/merge';
import session from '../../helpers/Session';

const internalFetch = (url, config = {}) => {
  const internalConfig = merge(
    {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${session.getToken()}`,
        'Content-Type': 'application/json'
      }
    },
    config
  );
  return fetch(url, internalConfig).then(async res => {
    if (res.status === 400 && (await res.json()).error === 'JWT token expired') {
      session.destroy();
    } else {
      return res;
    }
  });
};

export default internalFetch;
