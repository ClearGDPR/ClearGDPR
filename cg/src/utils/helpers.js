const Promise = require('bluebird');
const winston = require('winston');

const { VALID_RUN_MODES } = require('./constants');

function inControllerMode() {
  return process.env.MODE === VALID_RUN_MODES.CONTROLLER;
}

function getMyAddress() {
  return process.env.MY_ADDRESS;
}
function getTokenExpiry() {
  return process.env.JWT_TOKEN_EXPIRY || 604800;
}

const timeout = ms => new Promise(res => setTimeout(res, ms));

async function retryAsync(promise, options = {}) {
  let tries = 0;

  const maxRetries = options.maxRetries || 3;
  const interval = options.interval || 500;
  let delay = 0;

  while (tries < maxRetries) {
    await Promise.delay(delay);
    tries++;
    try {
      return await promise();
    } catch (e) {
      delay = interval * tries;
      winston.info(`Attempt no. ${tries} failed. Next attempt in ${delay}ms`);
      winston.log('debug', `Error: ${e.toString()}`);
    }
  }

  winston.error(`Operation failed after ${tries} attempts`);
}

module.exports = {
  getTokenExpiry,
  inControllerMode,
  getMyAddress,
  timeout,
  retryAsync
};
