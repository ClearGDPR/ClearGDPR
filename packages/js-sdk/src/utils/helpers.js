import { merge } from 'lodash/object';

export function parseJSON(response) {
  const contentType = response.headers.get('Content-Type');

  if (contentType.indexOf('application/json') > -1) {
    return new Promise(resolve =>
      response.json().then(json => {
        return resolve({
          status: response.status,
          ok: response.ok,
          data: json
        });
      })
    );
  } else {
    return new Promise(resolve =>
      response.text().then(text => {
        return resolve({
          status: response.status,
          ok: response.ok,
          data: text
        });
      })
    );
  }
}

export function buildRequestBody(body) {
  return body ? JSON.stringify(body) : null;
}

/**
 * Merges objects values
 * @param {Object} options objects to be merged
 * @returns {Object} New merged object based on params
 */
export function mergeOptions(...objects) {
  return merge(...objects);
}
