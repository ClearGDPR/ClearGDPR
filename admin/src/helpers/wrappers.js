import React from 'react';
import PropTypes from 'prop-types';
import lodash from 'lodash';
import { UsersProvider } from 'containers/Users/UsersContext';
import { ProcessorsProvider } from 'containers/Processors/ProcessorsContext';
import { PanelProvider } from 'containers/MainLayout/PanelContext';
import { RectificationsProvider } from 'containers/Rectifications/RectificationsContext';
import { StatsProvider } from 'containers/Dashboard/StatsContext';

export const withWrapper = Wrapper => Component => {
  let wrapper = props => (
    <Wrapper>
      <Component {...props}>{props.children}</Component>
    </Wrapper>
  );
  wrapper.propTypes = {
    children: PropTypes.node
  };
  return wrapper;
};

export const withContextProviders = lodash.flow.apply(null, [
  withWrapper(UsersProvider),
  withWrapper(ProcessorsProvider),
  withWrapper(RectificationsProvider),
  withWrapper(PanelProvider),
  withWrapper(StatsProvider)
]);
