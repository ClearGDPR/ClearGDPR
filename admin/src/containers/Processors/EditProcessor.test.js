import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import * as TestUtils from 'tests/helpers/TestUtils';

import { EditProcessorContainer } from './EditProcessor';

const setupShallow = propOverrides => {
  const props = Object.assign(
    {
      updateProcessor: jest.fn().mockReturnValue(Promise.resolve()),
      closePanel: jest.fn(),
      isLoading: false,
      processor: {}
    },
    propOverrides
  );
  const component = shallow(<EditProcessorContainer {...props} />);

  return { props, component };
};

describe('(Container) Edit Processor', () => {
  it('should render correctly when default props provided', async () => {
    const { component } = setupShallow();
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should populate errors on submit when update rejects with error', async () => {
    const { props, component } = setupShallow();
    const stubProcessor = {
      name: null,
      logoUrl: null,
      scopes: []
    };

    props.updateProcessor.mockReturnValue(Promise.reject(new Error('Test error')));
    component.instance().onSubmit(stubProcessor);

    await TestUtils.flushPromises();
    expect(component.state()).toMatchSnapshot();
  });

  it('should close panel when update succeeded', async () => {
    const { props, component } = setupShallow();
    const stubProcessor = {
      name: 'test.processor',
      logoUrl: 'https://this.is.a.mock.image.com/url.png',
      scopes: []
    };

    component
      .instance()
      .onSubmit(stubProcessor);

    await TestUtils.flushPromises();
    expect(props.closePanel).toHaveBeenCalled();
  });
});
