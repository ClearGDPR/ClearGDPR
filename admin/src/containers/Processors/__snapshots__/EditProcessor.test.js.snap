// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Container) Edit Processor should populate errors on submit when update rejects with error 1`] = `
Object {
  "errors": Object {
    "processor": "Error: Test error",
  },
}
`;

exports[`(Container) Edit Processor should render correctly when default props provided 1`] = `
<EditProcessorForm
  errors={Object {}}
  onSubmit={[Function]}
  values={Object {}}
/>
`;
